File: Editorial/103_Binary_Tree_Zigzag_Level_Order_Traversal/CSharp/1_BFS.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public IList<IList<int>> ZigzagLevelOrder(TreeNode root) {
        List<IList<int>> result = new List<IList<int>>();
        if (root == null)
            return result;
        Queue<TreeNode> nodeQueue = new Queue<TreeNode>();
        nodeQueue.Enqueue(root);
        nodeQueue.Enqueue(null);
        LinkedList<int> levelList = new LinkedList<int>();
        bool isOrderLeft = true;
        while (nodeQueue.Count > 0) {
            TreeNode currentNode = nodeQueue.Dequeue();
            if (currentNode != null) {
                if (isOrderLeft)
                    levelList.AddLast(currentNode.val);
                else
                    levelList.AddFirst(currentNode.val);
                if (currentNode.left != null)
                    nodeQueue.Enqueue(currentNode.left);
                if (currentNode.right != null)
                    nodeQueue.Enqueue(currentNode.right);
            } else {
                result.Add(new List<int>(levelList));
                levelList.Clear();
                if (nodeQueue.Count > 0)
                    nodeQueue.Enqueue(null);
                isOrderLeft = !isOrderLeft;
            }
        }

        return result;
    }
}

File: Editorial/103_Binary_Tree_Zigzag_Level_Order_Traversal/CPP/1_BFS.cpp
// Definition for a binary tree node.
//  struct TreeNode {
//      int val;
//     TreeNode *left;
//     TreeNode *right;
//     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
// };
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        if (root == nullptr) return vector<vector<int>>();
        vector<vector<int>> result;
        deque<TreeNode*> node_queue;
        node_queue.push_back(root);
        node_queue.push_back(nullptr);  // Delimiter for a single level
        deque<int> level_list;
        bool is_order_left = true;
        while (!node_queue.empty()) {
            TreeNode* temp = node_queue.front();
            node_queue.pop_front();
            if (temp != nullptr) {
                if (is_order_left)
                    level_list.push_back(temp->val);
                else
                    level_list.push_front(temp->val);
                if (temp->left) node_queue.push_back(temp->left);
                if (temp->right) node_queue.push_back(temp->right);
            } else {
                // Add current level to result and prepare for the next level
                result.push_back(
                    vector<int>{level_list.begin(), level_list.end()});
                level_list.clear();
                if (!node_queue.empty()) node_queue.push_back(nullptr);
                is_order_left = !is_order_left;
            }
        }
        return result;
    }
};
